ext/standard/tests/array/array_modify_arg_in_place_optimization.phpt

```diff
@@ -0,0 +1,344 @@
+$listStr = array_slice($listStr, 0, 1, true);
+$mapInt = array_slice($mapInt, 0, 1, true);
+$mapObj = array_slice($mapObj, 0, 2, true);
+
+
+echo "*** array merge with unpacking ***\n";
+// TODO
+
+echo "*** array_diff family ***\n";
+// TODO array_diff family should be optimized - https://github.com/php/php-src/pull/11060#discussion_r1175022196
+
+echo "*** array_intersect family ***\n";
+$oneListInt = array_slice($listInt, 0, 1, true);
+$oneMapObj = array_slice($mapObj, 0, 1, true);
+
+$listInt[] = -1;
+$ptrBefore = zend_get_zval_ptr($listInt);
+$listInt = array_intersect($listInt, $oneListInt);
+$ptrAfter = zend_get_zval_ptr($listInt);
+var_dump($ptrAfter === $ptrBefore);
+
+$mapObj[] = -1;
+$ptrBefore = zend_get_zval_ptr($mapObj);
+$mapObj = array_intersect($mapObj, $oneMapObj);
+$ptrAfter = zend_get_zval_ptr($mapObj);
+var_dump($ptrAfter === $ptrBefore);
+
+$listInt[] = -1;
+$ptrBefore = zend_get_zval_ptr($listInt);
+$listInt = array_intersect_assoc($listInt, $oneListInt);
+$ptrAfter = zend_get_zval_ptr($listInt);
+var_dump($ptrAfter === $ptrBefore);
+
+$mapObj[] = -1;
+$ptrBefore = zend_get_zval_ptr($mapObj);
+$mapObj = array_intersect_assoc($mapObj, $oneMapObj);
+$ptrAfter = zend_get_zval_ptr($mapObj);
+var_dump($ptrAfter === $ptrBefore);
+
+$listInt[] = -1;
+$ptrBefore = zend_get_zval_ptr($listInt);
+$listInt = array_intersect_key($listInt, $oneListInt);
+$ptrAfter = zend_get_zval_ptr($listInt);
+var_dump($ptrAfter === $ptrBefore);
+
+$listInt[] = -1;
+$ptrBefore = zend_get_zval_ptr($listInt);
+$listInt = array_intersect_ukey($listInt, $oneListInt, fn ($k1, $k2) => $k1 <=> $k2);
+$ptrAfter = zend_get_zval_ptr($listInt);
+var_dump($ptrAfter === $ptrBefore);
+
+$mapObj[] = -1;
+$ptrBefore = zend_get_zval_ptr($mapObj);
+$mapObj = array_intersect_ukey($mapObj, $oneListInt, fn ($k1, $k2) => $k1 <=> $k2);
+$ptrAfter = zend_get_zval_ptr($mapObj);
+var_dump($ptrAfter === $ptrBefore);
+
+echo "*** array_unique ***\n";
+
+$listInt[] = end($listInt);
+$ptrBefore = zend_get_zval_ptr($listInt);
+$listInt = array_unique($listInt);
+$ptrAfter = zend_get_zval_ptr($listInt);
+var_dump($ptrAfter === $ptrBefore);
+
+$mapObj[] = end($mapObj);
+$ptrBefore = zend_get_zval_ptr($mapObj);
+$mapObj = array_unique($mapObj, SORT_REGULAR);
+$ptrAfter = zend_get_zval_ptr($mapObj);
+var_dump($ptrAfter === $ptrBefore);
+
+echo "*** array_replace ***\n";
+
+$listInt[] = end($listInt);
+$ptrBefore = zend_get_zval_ptr($listInt);
+$listInt = array_replace($listInt, ['*']);
+$ptrAfter = zend_get_zval_ptr($listInt);
+var_dump($ptrAfter === $ptrBefore);
+
+$mapObj[] = end($mapObj);
+$ptrBefore = zend_get_zval_ptr($mapObj);
+$mapObj = array_replace($mapObj, ['*']);
+$ptrAfter = zend_get_zval_ptr($mapObj);
+var_dump($ptrAfter === $ptrBefore);
+
+print_r($listInt);
+print_r($mapObj);
+
+?>
+--EXPECT--
+*** array_merge ***
+bool(true)
+bool(true)
+bool(true)
+bool(true)
+bool(true)
+bool(true)
+bool(true)
+bool(true)
+bool(true)
+bool(true)
+bool(true)
+bool(true)
+bool(false)
```

Comment:
Is `zend_get_zval_ptr` implemented correctly? It seems `$ptrAfter === $ptrBefore` is always true, which is not expected here.

Reply:
You need to be aware of how zvals and arrays work behind the scenes.
A zval contains some data and some type information. For arrays this type is IS_ARRAY, and its data is a _pointer_ to the array. When I do `$a = $b`, a and b are two different zvals, but if b is an array, then a will point to the same array because the pointer is copied (and the array refcount is increased by one).
Similarly if you have:
```php
$a = [];
$a = [1];
$a = [1, 2];
```
then the `zend_get_zval_ptr` for $a will always return the same pointer for all three lines, because you're always asking the pointer for the container variable a. What you actually want to ask the engine is the pointer to the array (e.g. Z_ARRVAL_P(zval_pointer_for_a_here)`. Then you'll see that each line will have a different array pointer, while having the same zval pointer.

Reply:
> Note that https://github.com/php/php-src/pull/11060 only introduces the optimisation for RC1 temporaries, not CVs. https://github.com/php/php-src/pull/11166 enables the optimisation for CVs.

@nielsdos `$listInt` in the test should have RC = 1 (excl. in the foreach test), but still the results are very different vs. what I would expect, can you please give me an example of php code where variable is "RC1 temporaries" and where "CV"?

Reply:
@mvorisek CVs are never RC1 when passed to a function. https://github.com/php/php-src/pull/11166 will change this but only work with opcache (and optimizations) enabled. TMP/VAR will be RC1 if allocated within a function and then returned, of when a new value is generated from a temporary expression (e.g. `[0 => 1] + [1 => 2]`). `range` is another way to generate a temporary array with RC1.

Reply:
So `array_merge` function in `$listInt = array_merge($listInt, []);` expr, when `$listInt` had RC1 originally, will not be optimized, as `array_merge` will receive RC2? If yes, then can the testing using the before/after PTR be done before #11166 is landed?

Reply:
> So array_merge function in $listInt = array_merge($listInt, []); expr, when $listInt had RC1 originally, will not be optimized, as array_merge will receive RC2?

Right

> If yes, then can the testing using the before/after PTR be done before https://github.com/php/php-src/pull/11166 is landed?

Sure, but that will only work in a limited context. E.g. pseudo main (root of the file) won't work, so you'll need to wrap it in a separate function. And as mentioned above, it will only work when opcache is enabled, as the optimizer otherwise doesn't run.

Reply:
@nielsdos would you be ok to add this test to your #11166 directly?

Reply:
Let's wait until that PR is merged and keep this one separate. Otherwise it will be a back and forth between us two which is a lot of overhead.